### PR TITLE
CI job: Run tests on ubuntu xenial

### PR DESCRIPTION
**Pull request by kbushgit**
Signed-off-by: Kostiantyn Bushko <kbushko@intellias.com>

### CODE REVIEW

File: docker/Dockerfile.ubuntu.xenial [hunk 1]
```diff
@@ -50,14 +45,17 @@ RUN apt-get update && apt-get -y install --no-install-suggests --no-install-reco
...
   python3-gi \
   python3-openssl \
   python3-pip \
   python3-venv \
   sqlite3 \
   wget \
-  zip
+  zip \
+  fiu-utils \
+  libfiu-dev
```
Comment (patrickvacek) [inline docker/Dockerfile.ubuntu.xenial]: Ideally, these should be sorted alphabetically.

File: docker/Dockerfile.ubuntu.xenial [hunk 2]
```diff
@@ -50,14 +45,17 @@ RUN apt-get update && apt-get -y install --no-install-suggests --no-install-reco
   opensc \
   pkg-config \
   psmisc \
+  python-dev \
```
Comment (patrickvacek) [inline docker/Dockerfile.ubuntu.xenial]: It's pretty annoying that we need python2 and python3. What's python2 required for?

File: docker/Dockerfile.ubuntu.xenial [hunk 3]
```diff
@@ -69,6 +67,23 @@ RUN apt-get update && apt-get -y install \
...
+RUN make -j4
+RUN make install
+
+WORKDIR /boost
+RUN wget https://sourceforge.net/projects/boost/files/boost/1.64.0/boost_1_64_0.tar.gz -nv
+RUN tar -xzf boost_1_64_0.tar.gz
+WORKDIR /boost/boost_1_64_0
+RUN ./bootstrap.sh --prefix=/usr/
+RUN ./b2
+RUN ./b2 install
```
Comment (patrickvacek) [inline docker/Dockerfile.ubuntu.xenial]: This is pretty unfortunate. What do we require from the newer versions?

File: src/libaktualizr/primary/CMakeLists.txt [hunk 4]
```diff
@@ -44,3 +44,7 @@ target_include_directories(t_uptane_key PUBLIC ${PROJECT_SOURCE_DIR}/src/virtual
 target_link_libraries(t_uptane_key virtual_secondary)
 
 aktualizr_source_file_checks(${SOURCES} ${HEADERS} ${TEST_SOURCES})
+
+set_target_properties (t_aktualizr PROPERTIES COMPILE_FLAGS -pthread LINK_FLAG -pthread)
+set_target_properties (t_aktualizr_fullostree PROPERTIES COMPILE_FLAGS -pthread LINK_FLAGS -pthread)
+set_target_properties (t_uptane_key PROPERTIES COMPILE_FLAGS -pthread LINK_FLAGS -pthread)
```
Comment (patrickvacek) [inline src/libaktualizr/primary/CMakeLists.txt]: I would prefer if we didn't have to have this hardcoded and duplicated for all these targets. Why is this required?

File: docker/Dockerfile.ubuntu.xenial (hunk 2 above)
Comment (kbushgit) [reply docker/Dockerfile.ubuntu.xenial]: Python2 is required to build boost but in the final PR I will change it to be possible to build boost with Python3

File: docker/Dockerfile.ubuntu.xenial [hunk 5]
```diff
@@ -69,6 +67,23 @@ RUN apt-get update && apt-get -y install \
   softhsm2 \
   valgrind
 
+
+WORKDIR /cmake
+RUN wget https://github.com/Kitware/CMake/releases/download/v3.14.4/cmake-3.14.4.tar.gz -nv
```
Comment (lbonn) [inline docker/Dockerfile.ubuntu.xenial]: We might be stuck with having to compile boost but it feels like it should be easier to live with an older CMake version. Do you know what's blocking us from being compatible with CMake on Xenial?

File: src/libaktualizr/primary/CMakeLists.txt (hunk 4 above)
Comment (kbushgit) [reply src/libaktualizr/primary/CMakeLists.txt]: We have a few targets which use libfiu, for these targets, we get linking error " undefined reference to pthread_atfork" I think it related to GCC version there are some differences in -lpthread and -pthread

File: docker/Dockerfile.ubuntu.xenial (hunk 5 above)
Comment (kbushgit) [reply docker/Dockerfile.ubuntu.xenial]: The CMake version lower than 3.8 does not properly support boost 1.64 or higher, it's related to Finding an installed library.

File: docker/Dockerfile.ubuntu.xenial (hunk 3 above)
Comment (kbushgit) [reply docker/Dockerfile.ubuntu.xenial]: As we discussed in chat currently we have a requirement boost 1.64 only for building tests, for building aktualizr Release we can use min boost version 1.58.
The question is, should we up minimum boost version for the project? 

The minimum possible boost version for us is 1.64

File: docker/Dockerfile.ubuntu.xenial (hunk 5 above)
Comment (lbonn) [reply docker/Dockerfile.ubuntu.xenial]: Ah gotcha, I forgot about that, makes sense.

File: docker/Dockerfile.ubuntu.xenial (hunk 3 above)
Comment (lbonn) [reply docker/Dockerfile.ubuntu.xenial]: Another solution could be to use an alternative to `boost::process` for our tests, maybe get something similar from a third party library. That takes a bit more dev effort but would simplify the CI part.

Comment (patrickvacek) [reply docker/Dockerfile.ubuntu.xenial]: `boost::process` is pretty nice though, just sayin'.

Comment (kbushgit) [reply docker/Dockerfile.ubuntu.xenial]: I believe it is possible to avoid using boost::process and replace it with analog system call for() exec() etc, but I think we should try to be up to date with the third-party libraries we use, if it's possible.
I believe that the boost::process is not a single feature in the boost 1.64 we can use in further development...

File: src/libaktualizr/primary/CMakeLists.txt (hunk 4 above)
Comment (kbushgit) [reply src/libaktualizr/primary/CMakeLists.txt]: I had installed GCC 7 on ubuntu xenial and the result is the same as for gcc 5, which is the default for the xenial, so the issue is not related to GCC version...

File: docker/Dockerfile.ubuntu.xenial (hunk 3 above)
Comment (patrickvacek) [reply docker/Dockerfile.ubuntu.xenial]: Sorry, I don't think I understand your comment. What are you proposing?

Comment (kbushgit) [reply docker/Dockerfile.ubuntu.xenial]: I think we should use boost 1.64 as a min version.

Comment (patrickvacek) [reply docker/Dockerfile.ubuntu.xenial]: Yeah, I think I agree. It's a bit of a pain, but undoing Mike's rather nice code using `boost::process` seems worse.

Comment (lbonn) [reply docker/Dockerfile.ubuntu.xenial]: Ok then we're settled I guess, sorry for derailing

Comment (patrickvacek) [reply docker/Dockerfile.ubuntu.xenial]: No problem, the discussion is still worthwhile, and I was the one who complained originally!

Comment (mike-sul) [reply docker/Dockerfile.ubuntu.xenial]: test_utils.h::Process (the one that relies on boost::process) can be potentially replaced by Utils::shell() that relies on stdio::popen() if it's really required.

File: ci/gitlab/.gitlab-ci.yml [hunk 6]
```diff
@@ -89,6 +90,28 @@ coverage:
     - aws s3 cp s3://ota-gitlab-ci/hereotaconnect_prod.zip $CI_PROJECT_DIR/credentials.zip
     - ./scripts/test.sh
 
+coverage-xenial:
+  variables:
+    TEST_BUILD_DIR: 'build-coverage-xenial'
+    TEST_CMAKE_BUILD_TYPE: 'Valgrind'
+    TEST_WITH_COVERAGE: '1'
+    TEST_WITH_P11: '1'
+    TEST_WITH_FAULT_INJECTION: '1'
```
Comment (patrickvacek) [inline ci/gitlab/.gitlab-ci.yml]: Let's try disabling the P11 and fault injection variables and then I think there's just a couple small things to finish up.

**Comment by lbonn** [comment]
I think the remaining failure comes from a change in Openssl's output between 1.0 and 1.1. Should be possible to work around.

**Comment by mike-sul** [comment]
> I think the remaining failure comes from a change in Openssl's output between 1.0 and 1.1. Should be possible to work around.

It most likely my fault since I introduced the test(s) that depends on specific openssl version output.
I can fix it.

File: docker/Dockerfile.ubuntu.xenial (hunk 2 above)
Comment (lbonn) [reply docker/Dockerfile.ubuntu.xenial]: Is this still part of the plan?

File: src/libaktualizr/primary/CMakeLists.txt [hunk 7]
```diff
@@ -1,3 +1,7 @@
+if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+    set(CMAKE_CXX_FLAGS -pthread)
```
Comment (lbonn) [inline src/libaktualizr/primary/CMakeLists.txt]: Does it also work if only added in the root CMakeLists.txt, instead of having to duplicate that in several places?

Comment (kbushgit) [reply src/libaktualizr/primary/CMakeLists.txt]: Probably Yes, the global flag should work for all targets, I'll check it...

File: docker/Dockerfile.ubuntu.xenial (hunk 2 above)
Comment (kbushgit) [reply docker/Dockerfile.ubuntu.xenial]: No, this needed only for building boost with enabled boost::python module, we do not use it right now, so it can be removed

File: CMakeLists.txt [hunk 8]
```diff
@@ -7,6 +7,8 @@ cmake_policy(SET CMP0009 NEW)
 
 project(aktualizr)
 
+include_directories(/usr/include/ostree-1)
```
Comment (lbonn) [inline CMakeLists.txt]: Is this needed? We have this line later: `include_directories(${LIBOSTREE_INCLUDE_DIRS})`

File: docker/Dockerfile-test-install.ubuntu.xenial [hunk 9]
```diff
@@ -23,4 +18,9 @@ RUN apt-get update && apt-get install -y --no-install-suggests --no-install-reco
   python3 \
   zip
 
+RUN apt install -y software-properties-common
+RUN add-apt-repository ppa:mhier/libboost-latest
+RUN apt update
+RUN apt install -y libboost1.67
```
Comment (lbonn) [inline docker/Dockerfile-test-install.ubuntu.xenial]: Why is this needed for test-install? We install the package built against stock libboost in this case.

File: docker/Dockerfile.ubuntu.xenial [hunk 10]
```diff
@@ -13,26 +13,23 @@ RUN apt-get update && apt-get -y install --no-install-suggests --no-install-reco
   autoconf \
   asn1c \
   automake \
+  awscli \
   bison \
   ccache \
-  cmake \
+  clang-format-6.0 \
```
Comment (lbonn) [inline docker/Dockerfile.ubuntu.xenial]: Do we need clang-format here?

File: docker/Dockerfile-test-install.ubuntu.xenial (hunk 9 above)
Comment (kbushgit) [reply docker/Dockerfile-test-install.ubuntu.xenial]: Boost shared library must be installed on system where we run aktualizr, this is installation from alternative repository which has version of the boost 1.67, the default repo for xenial allows to install only 1.58

Comment (lbonn) [reply docker/Dockerfile-test-install.ubuntu.xenial]: Yes but that's what gets compiled on the actual debian package and we probably want to keep it like that. It doesn't make sense to provide a xenial deb package and ask people to install a crazy new boost version.

File: docker/Dockerfile.ubuntu.xenial (hunk 10 above)
Comment (kbushgit) [reply docker/Dockerfile.ubuntu.xenial]: maybe this is for the case when we decide to check the formatting, not sure...

File: docker/Dockerfile-test-install.ubuntu.xenial (hunk 9 above)
Comment (kbushgit) [reply docker/Dockerfile-test-install.ubuntu.xenial]: Here we back to the question which is the minimum boost library we should use! 

From boost FAQ
How can the Boost libraries be used successfully for important projects? Many of the Boost libraries are actively maintained and improved, so backward compatibility with prior version isn't always possible. Deal with this by freezing the version of the Boost libraries used by your project. Only upgrade at points in your project's life cycle where a bit of change will not cause problems. Individual bug fixes can always be obtained from the boost repository.

https://www.boost.org/users/faq.html

File: CMakeLists.txt (hunk 8 above)
Comment (kbushgit) [reply CMakeLists.txt]: Please check this cmake module
https://github.com/advancedtelematic/aktualizr/blob/78e19207414a2373d70d63ae5a4810f9b956cbd8/cmake-modules/FindOSTree.cmake#L1

Normally we install ostree in default path "/usr" so the include dir will be located on /usr/include/ostree-1 and lib in /usr/lib!

Not sure this module can properly find the include directories since it nested to ostree-1, for *.so file it works because it located in default path for the linker

Comment (kbushgit) [reply CMakeLists.txt]: That's what I getting!

In file included from /home/kbushko/work/pr/feat_OTA-2753_CI_ubuntu_xenial_build_tests/aktualizr/tests/aktualizr_cycle_simple.cc:10:0:
/home/kbushko/work/pr/feat_OTA-2753_CI_ubuntu_xenial_build_tests/aktualizr/src/libaktualizr/package_manager/ostreemanager.h:8:10: fatal error: ostree.h: No such file or directory
 #include <ostree.h>
          ^~~~~~~~~~

File: docker/Dockerfile-test-install.ubuntu.xenial (hunk 9 above)
Comment (lbonn) [reply docker/Dockerfile-test-install.ubuntu.xenial]: Ah yes I just realized that we also build the xenial package against this new boost then... That makes this .deb package quite useless. It doesn't make sense to say we support xenial if we recompile most of our dependencies there.
And we don't even need a new boost for the application, only the tests!

File: CMakeLists.txt (hunk 8 above)
Comment (kbushgit) [reply CMakeLists.txt]: Ah, this is not an issue with FindOSTree.cmake, the problem is in the aktualizr_cycle_simpl.cc this test include ostreemanager.h which includes ostree.h and if we build project without OSTREE it will fail since find_package for OSTree is not executed!

This actually is a case when TEST which executed by default has dependencies on the package which is optional, in this particular case -DBUILD_OSTREE

**Comment by lbonn** [comment]
As we've agreed this morning, we'll close this PR as we've already started opening others for the bits we want to keep.
